pandas/core/groupby/groupby.py

```diff
@@ -718,12 +718,27 @@ def _set_result_index_ordered(
 
         # the values/counts are repeated according to the group index
         # shortcut if we have an already ordered grouper
-        if not self.grouper.is_monotonic:
-            index = Index(np.concatenate(self._get_indices(self.grouper.result_index)))
-            result.set_axis(index, axis=self.axis, inplace=True)
-            result = result.sort_index(axis=self.axis)
+        if self.grouper.is_monotonic:
+            result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
+            return result
+
+        # row order is scrambled => sort the rows by position in original index
+        original_positions = Index(
+            np.concatenate(self._get_indices(self.grouper.result_index))
+        )
+        result.set_axis(original_positions, axis=self.axis, inplace=True)
+        result = result.sort_index(axis=self.axis)
+
+        rows_dropped = len(result.index) < len(self._selected_obj)
+
+        # get index by slicing original index according to original positions
+        # slice drops attrs => use set_axis when no rows were dropped
+        if rows_dropped:
+            sorted_indexer = result.index
```

Comment:
can we just *always* use this formulation? e.g. just use the if clause regardless here?

Reply:
well I mean for axis==0; also I think this is incorrect for axis==1 (though it likley won't trigger the if clause).

can you refactor to do this?

Reply:
> can we just _always_ use this formulation? e.g. just use the if clause regardless here?

We can but the slice will drop some attributes (for example frequency for a datetime index - we have testcase which won't quite roundtrip if we do this). Will change as long as you're ok with that

Reply:
> well I mean for axis==0; also I think this is incorrect for axis==1 (though it likley won't trigger the if clause).

Right because for `axis=1` we don't drop rows.

FTR we don't fully support/test the `axis=1` case atm. I'll do that in a follow-on once this is in

Reply:
> well I mean for axis==0; also I think this is incorrect for axis==1 (though it likley won't trigger the if clause).
> 
> can you refactor to do this?

On second thought I don't think we can 

AFAIK if we don't drop rows then we keep the original index which isn't an indexer into the original object's index. When we drop rows we reindex with an integer index (and that's a valid argument to `__getitem__`). So the case of no rows dropped doesn't fall through the slice and I think we do have to treat those two cases separately here